src/main/kotlin/com/hypto/iam/server/service/UserAuthService.kt

```diff
@@ -1,19 +1,62 @@
 package com.hypto.iam.server.service
 
+import com.hypto.iam.server.authProviders.AuthProviderRegistry
 import com.hypto.iam.server.db.repositories.OrganizationRepo
 import com.hypto.iam.server.db.repositories.UserAuthRepo
+import com.hypto.iam.server.db.repositories.UserRepo
 import com.hypto.iam.server.exceptions.EntityNotFoundException
+import com.hypto.iam.server.models.BaseSuccessResponse
 import com.hypto.iam.server.models.UserAuthMethod
 import com.hypto.iam.server.models.UserAuthMethodsResponse
+import com.hypto.iam.server.security.AuthMetadata
+import com.hypto.iam.server.security.AuthenticationException
+import com.hypto.iam.server.security.TokenCredential
+import com.hypto.iam.server.security.TokenType
+import com.hypto.iam.server.security.UserPrincipal
 import com.hypto.iam.server.utils.IamResources
 import com.hypto.iam.server.utils.ResourceHrn
 import org.koin.core.component.KoinComponent
 import org.koin.core.component.inject
 
 class UserAuthServiceImpl : KoinComponent, UserAuthService {
     private val organizationRepo: OrganizationRepo by inject()
+    private val userRepo: UserRepo by inject()
     private val userAuthRepo: UserAuthRepo by inject()
 
+    override suspend fun createUserAuth(
+        organizationId: String,
+        userId: String,
+        issuer: String,
+        token: String,
+        principal: UserPrincipal,
+    ): BaseSuccessResponse {
+        if (issuer != TokenServiceImpl.ISSUER) {
+            val authProvider =
+                AuthProviderRegistry.getProvider(issuer) ?: throw AuthenticationException(
+                    "Invalid issuer",
+                )
+            val oAuthUserPrincipal = authProvider.getProfileDetails(TokenCredential(token, TokenType.OAUTH))
+            val user =
+                userRepo.findByEmail(oAuthUserPrincipal.email) ?: throw AuthenticationException(
+                    "User has not signed up yet",
+                )
+            if (principal.hrnStr != user.hrn) {
+                throw AuthenticationException("Can't add auth method for another user")
+            }
+            val userAuth = userAuthRepo.fetchByUserHrnAndProviderName(user.hrn, issuer)
+            if (userAuth == null) {
+                userAuthRepo.create(
+                    user.hrn,
+                    oAuthUserPrincipal.issuer,
+                    oAuthUserPrincipal.metadata?.let { AuthMetadata.toJsonB(it) },
+                )
+            } else {
+                throw AuthenticationException("User already has this auth method")
```

Comment:
Why is this an `AuthenticationException`?
Should this simply return `BaseSuccessResponse(true)`?